### ISSUE TITLE
Forwarding does not work properly on inline m2m autocomplete fields

### ISSUE THREAD

**Issue by meesterguyman** [yourlabs/django-autocomplete-light#713]
On a single admin page, I have a foreignkey autocomplete, an m2m autocomplete, and an m2m autocomplete on an inline form. All return results, but the m2m on the inline form does not pass along the forwarded field data in the requests as the other two do, so results are not what they should be. I have not tested a foreignkey autocomplete on an inline yet, but this may be an issue for inlines in general.

**Comment by jpic** [comment]
Could you try to reproduce this in a fork of test_project please ? Itll
help understanding a lot

**Comment by meesterguyman** [comment]
Will see what I can do.

FYI, I did attempt this on a foreignkey autocomplete in an inline, and this doesn't work either. So the issue has to do with forwards not being included in the ajax request, whenever the field is in an inline.

For example, when I add forwards to a form, I might see something like the following on the field that I added it to:

data-autocomplete-light-forward="field1,field2"

I am seeing this being added as expected on all of the fields. The problem appears to be with the javascript that goes and grabs the values of these fields, just prior to the request.

On non-inline forms, the request looks something like this, depending on the fowards:

/my-autocomplete/?q=blahblah&forward=%7Bfield1%3A%201%2C%20field2%3A%202%7D

On inline forms, the request always looks like this, regardless of the forwards registered to the field:

/my-autocomplete/?q=blahblah&forward=%7B%7D

The javascript is just failing to get the field values for some reason for the inlines. That may be enough to troubleshoot. If not, I'll try to get something up later.

**Comment by jpic** [comment]
I couldn't reproduce your issue [in the test project](http://dal-yourlabs.rhcloud.com/admin/linked_data/testmodel/add/), see the first XHR is without having set the owner and the second one after having set the owner to test:
![2016-07-22-134436_647x300_scrot](https://cloud.githubusercontent.com/assets/94636/17056344/fb23983a-5012-11e6-9d81-97874255d570.png)

Mainly [getFormPrefix()](https://github.com/yourlabs/django-autocomplete-light/blob/master/src/dal/static/autocomplete_light/autocomplete.init.js#L8-L32) is the JS function that tries to guess the form prefix for an HTML form field. So, perhaps it's not working correctly in your case ?

[add_forwards](https://github.com/yourlabs/django-autocomplete-light/blob/5c4ba9aca1b76a07f1915e75f901a9643f5de9d5/src/dal_select2/static/autocomplete_light/select2.js#L2-L16) (oops, need to rename this to addForwards) would be the JS function that guesses the data to forward, so that sounds like a good place to break your JS debugger on.

Let us know how it goes

**Comment by meesterguyman** [comment]
That's interesting. I am getting different results on the test project you sent. I am seeing "owner" in the request, which is more than what I see in our project, but I am not seeing its value when I use the test inline. I get the following with "test" owner selected from the inline:

?q=jh&forward=%7B%22owner%22%3A%22%22%7D

There should be a 1 in there, but there isn't. The one in other test autocomplete works fine, but the one in the inline is always blank.

**Comment by meesterguyman** [comment]
Okay, think I figured out what's going on. We are attempting to reference fields in the main form from our inline form. The add_forwards function appears to limit forwards to the inline form by its use of the inline form prefix. The fields we are referencing do not exist in the inline, so of course the forwards are ending up blank.

In our particular situation, our inline is being used to add media that is tagged to a particular company. The company is available in the main form, not in the inline. So there is a use case for being able to reference forward fields that are outside of the inline from within the inline form.

I would suggest that when there are forward fields within the inline, first attempt to look for the field within the inline using the inline form prefix. If no such element exists, then look for the field without the prefix as a fallback. That, or send both values over, if both happen to be found. Something along those lines would address our issues.

**Comment by meesterguyman** [comment]
Something like the following would work.

```
    function add_forwards(element) {
        var forward = element.attr('data-autocomplete-light-forward');
        if (forward !== undefined) {
            forward = forward.split(',');

            var prefix = $(element).getFormPrefix();
            var data_forward = {};

            for (var key in forward) {
                var $field = $('[name=' + prefix + forward[key] + ']');
                if (!$field.length)
                    $field = $('[name=' + forward[key] + ']');
                if ($field.length)
                    data_forward[forward[key]] = $field.val();
            }

            return JSON.stringify(data_forward);
        }
    }
```

**Comment by jpic** [comment]
Interresting, @gagarski what's your opinion about forwarding fields that don't have the prefix ?

**Comment by gagarski** [comment]
@jpic, the fix from @meesterguyman has been merged: 643374429547629b85f14c67ef7335a0153a0908

And it also should be working with new forward configuration format.

**Comment by jpic** [comment]
Awesome

Thanks @meesterguyman and @gagarski !
